core/src/main/java/org/apache/brooklyn/policy/InvokeEffectorOnSetChange.java

```diff
@@ -0,0 +1,203 @@
+            "onAdded",
+            "Name of the effector to invoke when entries are added to the collection.");
+
+    public static final ConfigKey<String> ON_REMOVED_EFFECTOR_NAME = ConfigKeys.newStringConfigKey(
+            "onRemoved",
+            "Name of the effector to invoke when entries are removed from the collection.");
+
+    public static final ConfigKey<String> PARAMETER_NAME = ConfigKeys.newStringConfigKey(
+            "parameterName",
+            "The name of the parameter to supply to the effectors",
+            "value");
+
+    /** The previous version of the set against which events will be compared. */
+    private Set<Object> previous = Collections.emptySet();
+
+    /** Guards accesses of previous. */
+    private final Object[] updateLock = new Object[0];
+
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        Sensor<? extends Collection<?>> sensor =
+                checkNotNull(getConfig(TRIGGER_SENSOR), "Value required for " + TRIGGER_SENSOR.getName());
+
+        checkArgument(Strings.isNonBlank(getConfig(PARAMETER_NAME)), "Value required for " + PARAMETER_NAME.getName());
+
+        // Fail straight away if neither effector is found.
+        if (getEffector(getOnAddedEffector()).isAbsentOrNull() &&
+                getEffector(getOnRemovedEffector()).isAbsentOrNull()) {
+            throw new IllegalArgumentException("Value required for one or both of " + ON_ADDED_EFFECTOR_NAME.getName() +
+                    " and " + ON_REMOVED_EFFECTOR_NAME.getName());
+        }
+
+        // Initialise `present` before subscribing.
+        Collection<?> current = entity.sensors().get(getTriggerSensor());
+        synchronized (updateLock) {
+            previous = (current != null) ? new HashSet<>(current) : Collections.emptySet();
+        }
+        subscriptions().subscribe(entity, sensor, this);
+    }
+
+    @Override
+    public void onEvent(SensorEvent<Collection<?>> event) {
+        final Set<Object> added = new HashSet<>(), removed = new HashSet<>();
+        Set<Object> newValue = new HashSet<>(event.getValue());
+        synchronized (updateLock) {
+            // Not using .immutableCopy() in case either set contains `null`.
+            Sets.difference(newValue, previous).copyInto(added);
+            Sets.difference(previous, newValue).copyInto(removed);
+
+            // Only perform an update if the collections have changed.
+            if (!added.isEmpty() || !removed.isEmpty()) {
+                for (Object o : added) {
+                    onAdded(o);
+                }
+                for (Object o : removed) {
+                    onRemoved(o);
+                }
+                entity.sensors().set(getPreviousSensor(), previous);
+                requestPersist();
+                // In advance of the next event.
+                previous = Collections.unmodifiableSet(newValue);
```

Comment:
Just to be clear - do you mean to ask whether we really need a separate _field_? Rather than sensor. Assuming so, yes, the previous sensor is just for users to look at. I thought we'd decided that was desirable previously. I don't have any strong feelings about the `previous` sensor - am happy enough to get rid of it but can see that it might be useful.